### PR TITLE
Complete developer-test project

### PR DESCRIPTION
**Pull request by manytostao-zz**
Almost every request was covered including the use of Angular Material styles.
Tests were written in server project only. None was included in client project due to issues with drivers for browses (Chrome, specifically).

### CODE REVIEW

File: bookstore/client/src/app/_services/crud.service.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import {Http} from '@angular/http';
+import {Observable} from 'rxjs/Observable';
+
+@Injectable()
+export abstract class CrudService {
+  protected url = 'http://localhost:3000/';
+
+  protected model = '';
+
+  protected constructor(protected http: Http) {
```
Comment (carlososiel) [inline bookstore/client/src/app/_services/crud.service.ts:11]: No is recommended used Http, is deprecated, the recommended is to use HttpClient

File: bookstore/client/src/app/_components/confirmation-dialog/confirmation-dialog.component.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import {Component, Inject, OnInit} from '@angular/core';
+import {MAT_DIALOG_DATA, MatDialogRef} from '@angular/material';
+import {BookEditComponent} from '../../book/book-edit/book-edit.component';
```
Comment (carlososiel) [inline bookstore/client/src/app/_components/confirmation-dialog/confirmation-dialog.component.ts:3]: Angular style say that need a blank line to separate import blocks

File: bookstore/client/src/app/_components/home/home.component.html [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<!--The content below is only a placeholder and can be replaced.-->
+<div style="text-align:center">
```
Comment (carlososiel) [inline bookstore/client/src/app/_components/home/home.component.html:2]: No is recommended to used inline styles

File: bookstore/client/src/app/_components/nav-bar/nav-bar.component.html [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<nav>
+  <mat-toolbar color="primary">
+    <span>Bookstore</span>
+    <span style="flex: 1 1 auto"></span>
```
Comment (carlososiel) [inline bookstore/client/src/app/_components/nav-bar/nav-bar.component.html:4]: No is recommended use inline styles

File: bookstore/client/src/app/_services/crud.service.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import {Injectable} from '@angular/core';
+import {Http} from '@angular/http';
+import {Observable} from 'rxjs/Observable';
+
+@Injectable()
+export abstract class CrudService {
+  protected url = 'http://localhost:3000/';
```
Comment (carlososiel) [inline bookstore/client/src/app/_services/crud.service.ts:7]: These part can gone from environment files

File: bookstore/client/src/app/_services/crud.service.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  delete(id: string) {
+    return this.http.delete(this.url + this.model + '/' + id);
+  }
+
+  save(model: any) {
+    return this.http.post(this.url + this.model, model);
+  }
+
+  getDisplayedColumns(): string[] {
+    return undefined;
```
Comment (carlososiel) [inline bookstore/client/src/app/_services/crud.service.ts:35]: return undefined?

File: bookstore/client/src/app/app.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { TestBed, async } from '@angular/core/testing';
+import { AppComponent } from './app.component';
```
Comment (carlososiel) [inline bookstore/client/src/app/app.component.spec.ts:2]: Angular style say that need a blank line to separate import blocks

File: bookstore/client/src/app/app.module.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import {BrowserModule} from '@angular/platform-browser';
+import {BrowserAnimationsModule} from '@angular/platform-browser/animations';
+import {NgModule} from '@angular/core';
+
+
```
Comment (carlososiel) [inline bookstore/client/src/app/app.module.ts:5]: Only 1 blank line

File: bookstore/client/src/app/author/author-crud.service.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import {CrudService} from '../_services/crud.service';
```
Comment (carlososiel) [inline bookstore/client/src/app/author/author-crud.service.ts:1]: Angular style say that the correct order for import are: 1- Angular core, 2-Third part libraries and finally local imports, all import blocks separate with a blank line

File: bookstore/client/src/app/author/author-crud.service.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+import {Http} from '@angular/http';
+import {Injectable} from '@angular/core';
+
+@Injectable()
+export class AuthorCrudService extends CrudService {
+
+  constructor(http: Http) {
```
Comment (carlososiel) [inline bookstore/client/src/app/author/author-crud.service.ts:8]: Http is deprecated, instead of use HttpClient

**Review by carlososiel — COMMENTED** [review]
I review the frontend part in the PR, I have some recommendations that you need  have in next times:
- Used of Http instead of HttpClient, the first are deprecated
- Angular Style: Import order and separation between blocks
- Use Reactive Forms or something similar to save, edit data and add validation for fields
- Not use inline styles in html
- Use environements file instead of hard code values for urls
- Not use console.log or something similar for show errors, you can use MatSnackBar from material library

File: bookstore/client/src/app/category/category-crud.service.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import {CrudService} from '../_services/crud.service';
+import {Http} from '@angular/http';
+import {Injectable} from '@angular/core';
+
+@Injectable()
+export class CategoryCrudService extends CrudService {
+
+  constructor(http: Http) {
```
Comment (jcuban90) [inline bookstore/client/src/app/category/category-crud.service.ts:8]: Http deprecated use HttpClient

File: bookstore/client/src/app/category/category-list/category-list.component.ts [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+@Component({
+  selector: 'app-category-list',
+  templateUrl: './category-list.component.html',
+  styleUrls: ['./category-list.component.css'],
+  providers: [{provide: CrudService, useClass: CategoryCrudService}]
+})
+export class CategoryListComponent implements OnInit {
+
+  displayedColumns = [];
```
Comment (jcuban90) [inline bookstore/client/src/app/category/category-list/category-list.component.ts:17]: add data type

**Comment by manytostao** [comment]
I understand all recommendations made on this pull request, all easily solvable. What I don't understand though is, do I have to implement them in order to be ellegible for a position in your company or are they just as general culture?

**Comment by manytostao-zz** [comment]
I understand all recommendations made on this pull request, all easily solvable. What I don't understand though is, do I have to implement them in order to be ellegible for a position in your team or are they just as general culture?

**Pull request closed without merging**